From e91921ffe0f8306b2f3ef8a4e34eac085936e659 Mon Sep 17 00:00:00 2001
From: Girish S G <girishsg@ti.com>
Date: Tue, 8 Mar 2011 20:15:45 -0600
Subject: OMAP4:PM:TWL: TWL6030 register settings to prevent power leakage

Miscellaneous settings on TWL6030 to save power. This is done one
time during twl init. Drivers to take care for active/inactive usecase.

REVISIT.

Change-Id: I8fdab9736d1ef86444ff74983cc50a264887f1ad
Signed-off-by: Girish S G <girishsg@ti.com>
---
 drivers/mfd/twl-core.c  |   38 ++++++++++++++++++++++++++++++++++++++
 include/linux/i2c/twl.h |    1 +
 2 files changed, 39 insertions(+), 0 deletions(-)

diff --git a/drivers/mfd/twl-core.c b/drivers/mfd/twl-core.c
index 202e2ca..91222cf 100644
--- a/drivers/mfd/twl-core.c
+++ b/drivers/mfd/twl-core.c
@@ -992,6 +992,40 @@ static int __devexit twl_remove(struct i2c_client *client)
 	return 0;
 }
 
+static void _init_twl6030_settings(void)
+{
+	/* unmask PREQ transition */
+	twl_i2c_write_u8(TWL6030_MODULE_ID0, 0xE0, 0x02);
+
+	/* USB_VBUS_CTRL_CLR */
+	twl_i2c_write_u8(TWL6030_MODULE_ID1, 0xFF, 0x05);
+	/* USB_ID_CRTL_CLR */
+	twl_i2c_write_u8(TWL6030_MODULE_ID1, 0xFF, 0x07);
+
+	/* GPADC_CTRL */
+	twl_i2c_write_u8(TWL6030_MODULE_ID1, 0x00, 0x2E);
+	/* TOGGLE1 */
+	twl_i2c_write_u8(TWL6030_MODULE_ID1, 0x51, 0x90);
+	/* MISC1 */
+	twl_i2c_write_u8(TWL6030_MODULE_ID0, 0x00, 0xE4);
+	/* MISC2 */
+	twl_i2c_write_u8(TWL6030_MODULE_ID0, 0x00, 0xE5);
+
+	/*
+	 * BBSPOR_CFG - Disable BB charging. It should be
+	 * taken care by proper driver
+	 */
+	twl_i2c_write_u8(TWL6030_MODULE_ID0, 0x62, 0xE6);
+	/* CFG_INPUT_PUPD2 */
+	twl_i2c_write_u8(TWL6030_MODULE_ID0, 0x65, 0xF1);
+	/* CFG_INPUT_PUPD4 */
+	twl_i2c_write_u8(TWL6030_MODULE_ID0, 0x00, 0xF3);
+	/* CFG_LDO_PD2 */
+	twl_i2c_write_u8(TWL6030_MODULE_ID0, 0x00, 0xF5);
+	/* CHARGERUSB_CTRL3 */
+	twl_i2c_write_u8(TWL6030_MODULE_ID1, 0x21, 0xEA);
+}
+
 /* NOTE:  this driver only handles a single twl4030/tps659x0 chip */
 static int __devinit
 twl_probe(struct i2c_client *client, const struct i2c_device_id *id)
@@ -1089,6 +1123,10 @@ twl_probe(struct i2c_client *client, const struct i2c_device_id *id)
 	if (twl_class_is_6030())
 		twl_i2c_write_u8(TWL6030_MODULE_ID0, 0xE1, CLK32KG_CFG_STATE);
 
+	/* Remove unwanted settings on twl chip as part of twl init. */
+	if (twl_class_is_6030())
+		_init_twl6030_settings();
+
 	status = add_children(pdata, id->driver_data);
 fail:
 	if (status < 0)
diff --git a/include/linux/i2c/twl.h b/include/linux/i2c/twl.h
index eef4c17..e85364d 100644
--- a/include/linux/i2c/twl.h
+++ b/include/linux/i2c/twl.h
@@ -169,6 +169,7 @@
 
 #define TWL4030_CLASS_ID 		0x4030
 #define TWL6030_CLASS_ID 		0x6030
+
 unsigned int twl_rev(void);
 #define GET_TWL_REV (twl_rev())
 #define TWL_CLASS_IS(class, id)			\
-- 
1.6.0.4

